sonar-plugin-api/src/test/java/org/sonar/api/server/rule/internal/DefaultNewRuleTest.java

```diff
@@ -261,4 +272,68 @@ public void fail_if_trying_to_insert_context_aware_section_and_then_non_context_
+    rule.addDescriptionSection(createSectionWithContext(RESOURCES_SECTION_KEY, "ctx2"));
+    rule.addDescriptionSection(createSectionWithContext(HOW_TO_FIX_SECTION_KEY, "ctx1"));
+    rule.addDescriptionSection(createSectionWithContext(HOW_TO_FIX_SECTION_KEY, "ctx2"));
+    rule.validate();
+  }
+
+  @Test
+  public void validate_succeeds_when_single_contextualized_section_not_mixing_different_context_keys() {
+    rule.setHtmlDescription(RandomStringUtils.randomAlphanumeric(500));
+    rule.addDescriptionSection(RULE_DESCRIPTION_SECTION);
+    rule.addDescriptionSection(createSectionWithContext(HOW_TO_FIX_SECTION_KEY, "ctx1"));
+    rule.addDescriptionSection(createSectionWithContext(HOW_TO_FIX_SECTION_KEY, "ctx2"));
+    rule.validate();
+  }
+
+  @Test
+  public void validate_succeeds_when_no_contextualized_section() {
+    rule.setHtmlDescription(RandomStringUtils.randomAlphanumeric(500));
+    rule.addDescriptionSection(RULE_DESCRIPTION_SECTION);
+    rule.addDescriptionSection(createSectionWithContext(HOW_TO_FIX_SECTION_KEY, "ctx1"));
+    rule.addDescriptionSection(createSectionWithContext(HOW_TO_FIX_SECTION_KEY, "ctx2"));
+    rule.validate();
+  }
+
+  @Test
+  public void validate_fails_when_mixing_different_context_keys() {
+    rule.setHtmlDescription(RandomStringUtils.randomAlphanumeric(500));
+    rule.addDescriptionSection(createSectionWithContext(RESOURCES_SECTION_KEY, "ctx1"));
+    rule.addDescriptionSection(createSectionWithContext(RESOURCES_SECTION_KEY, "ctx2"));
+    rule.addDescriptionSection(createSectionWithContext(HOW_TO_FIX_SECTION_KEY, "ctx1"));
+    rule.addDescriptionSection(createSectionWithContext(HOW_TO_FIX_SECTION_KEY, "ctx3"));
+
+    assertThatThrownBy(rule::validate)
+      .hasMessage(MIXTURE_OF_CONTEXT_KEYS_BETWEEN_SECTIONS_ERROR_MESSAGE, HOW_TO_FIX_SECTION_KEY, "[ctx3, ctx1]", RESOURCES_SECTION_KEY, "[ctx1, ctx2]")
+      .isInstanceOf(IllegalArgumentException.class);
+  }
+
+  @Test
+  public void validate_fails_when_more_contexts_available_for_one_section() {
+    rule.setHtmlDescription(RandomStringUtils.randomAlphanumeric(500));
+    rule.addDescriptionSection(createSectionWithContext(RESOURCES_SECTION_KEY, "ctx1"));
+    rule.addDescriptionSection(createSectionWithContext(HOW_TO_FIX_SECTION_KEY, "ctx1"));
+    rule.addDescriptionSection(createSectionWithContext(HOW_TO_FIX_SECTION_KEY, "ctx2"));
+
+    assertThatThrownBy(rule::validate)
+      .hasMessage(MIXTURE_OF_CONTEXT_KEYS_BETWEEN_SECTIONS_ERROR_MESSAGE, HOW_TO_FIX_SECTION_KEY, "[ctx1, ctx2]", RESOURCES_SECTION_KEY, "[ctx1]")
+      .isInstanceOf(IllegalArgumentException.class);
+  }
+
+  private static RuleDescriptionSection createSectionWithContext(String sectionKey, String contextKey) {
+    return RuleDescriptionSection.builder()
+      .sectionKey(sectionKey)
+      .htmlContent(RandomStringUtils.randomAlphabetic(100))
+      .context(new Context(contextKey,contextKey + RandomStringUtils.randomAlphanumeric(10)))
```

Comment:
There is a slight formatting issue here with the comma + we can use static import for randomAlphanumeric
```suggestion
      .context(new Context(contextKey, contextKey + randomAlphanumeric(10)))
```